FileGetContentsLoader.php

```diff
@@ -27,95 +27,132 @@ public function loadDocument($url)
+                    $value = substr($http_response_header[$i], 5);
+                    $linkHeaderValues = array_merge($linkHeaderValues, explode(',', $value));
+                }
+            }
+        }
+
+        // Uh-oh - schema.org does no longer supports content negotiation!
+        // This re-routes the request to an appropriate resource as specified by the 'Link' header.
+        $base = new IRI($url);
+
+        foreach ($linkHeaderValues as $value) {
+            if (preg_match("/<(.[^>]+)>;/", $value, $uri)) {
+                $iri = new IRI($uri[1]);
 
-            $httpHeadersOffset = 0;
+                $link = array('uri' => $iri->isAbsolute() ? $uri[1] : (string) $base->resolve($iri));
 
-            stream_context_set_params($context, array('notification' =>
-                function ($code, $severity, $msg, $msgCode, $bytesTx, $bytesMax) use (
-                    &$remoteDocument, &$http_response_header, &$httpHeadersOffset
-                ) {
-                    if ($code === STREAM_NOTIFY_MIME_TYPE_IS) {
-                        $remoteDocument->mediaType = $msg;
-                    } elseif ($code === STREAM_NOTIFY_REDIRECTED) {
-                        $remoteDocument->documentUrl = $msg;
-                        $remoteDocument->mediaType = null;
+                preg_match_all("/;\s?([A-z][^,=]+)=\"(.[^\"]+)\"/", $value, $parameters);
 
-                        $httpHeadersOffset = count($http_response_header);
+                if (count($parameters) == 3) {
+                    $keys = $parameters[1];
+                    $values = $parameters[2];
+
+                    for ($i=0; $i < count($keys); $i++) {
+                        $link[trim($keys[$i])] = trim($values[$i]);
                     }
                 }
-            ));
 
-            if (false === ($input = @file_get_contents($url, false, $context))) {
-                throw new JsonLdException(
-                    JsonLdException::LOADING_DOCUMENT_FAILED,
-                    sprintf('Unable to load the remote document "%s".', $url),
-                    $http_response_header
-                );
-            }
-
-            // Extract HTTP Link headers
-            $linkHeaderValues = array();
-            if (is_array($http_response_header)) {
-                for ($i = count($http_response_header) - 1; $i > $httpHeadersOffset; $i--) {
-                    if (0 === substr_compare($http_response_header[$i], 'Link:', 0, 5, true)) {
-                        $value = substr($http_response_header[$i], 5);
-                        $linkHeaderValues[] = $value;
+                if (isset($link['rel']) && isset($link['type']) && ($link['rel'] === 'alternate')) {
+                    switch($link['type']) {
+                        case 'application/ld+json':
+                        case 'application/json':
+                            return $this->retrieveRemoteDocument($link['uri']);
+                            break;
                     }
```

Comment:
Please generalize parseContextLinkHeaders() instead of duplicating most of its code and then please move the handling of the alternate link header to down to where we check the media type. This should only trigger if we received something other than a JSON-LD doc

Reply:
@lanthaler I'm very happy to take a look at this, but do you have any example headers I can use as text fixtures in this refactor? I avoided eding `parseContextLinkHeaders()` as I did not want to accidentally break existing functionality.

Reply:
Thanks. I'd suggest to reduce the number of changes to a minimum (see other comment) as this code has proven quite stable over the years and already does all the parsing you need.